docs/FAQ.md

```diff
@@ -288,6 +289,38 @@ For a refresher on Go's recommended workspace organization, see the ["How To Wri
 `dep init -gopath`, which falls back to network mode when the packages are not
 found in `GOPATH`. `dep ensure` doesn't work with projects in `GOPATH`.
 
+## How do I use `dep` with Docker?
+
+`dep ensure -vendor-only` creates the vendor folder from a valid `Gopkg.toml` and `Gopkg.lock` without checking for Go code.
+This is especially useful for builds inside docker utilizing cache layers.
+
+Sample dockerfile:
+
+    FROM golang:1.8 AS builder
+
+    RUN apt-get update && apt-get install -y unzip --no-install-recommends && \
+        apt-get autoremove -y && apt-get clean -y && \
+        wget -O dep.zip https://github.com/golang/dep/releases/download/v0.3.0/dep-linux-amd64.zip && \
+        echo '96c191251164b1404332793fb7d1e5d8de2641706b128bf8d65772363758f364  dep.zip' | sha256sum -c - && \
+        unzip -d /usr/bin dep.zip && rm dep.zip
+
+    RUN mkdir -p /go/src/github.com/***
+    WORKDIR /go/src/github.com/***
+
+    COPY Gopkg.toml Gopkg.lock ./
+
+    RUN dep ensure -vendor-only
+    COPY . .
+    RUN CGO_ENABLED=0 GOOS=linux go build -ldflags "-s -w" -a -installsuffix cgo -o *** 
```

Comment:
Let's add a comment with a link to more info on what black magic is being used, specifically `ldflags`, `-a`, and `installstuffix cgo`. This has a mix of recommendations floating around on the net and it's not clear what is strictly needed to run in docker, for example why did we omit the netgo workaround as well?

I'm not sure what the best reference is, but https://blog.docker.com/2016/09/docker-golang/, at least explains some of the concerns of running a go binary in a container.

Reply:
I generally use these flags, but if a project needs cgo support this will not work. 

`CGO_ENABLE=0` disables cgo so this will automagically use netgo :wink:. 

[Build options](https://golang.org/cmd/go/#hdr-Compile_packages_and_dependencies) related to the line are as follows:
```
-a
	force rebuilding of packages that are already up-to-date.
-installsuffix suffix
	a suffix to use in the name of the package installation directory,
	in order to keep output separate from default builds.
	If using the -race flag, the install suffix is automatically set to race
	or, if set explicitly, has _race appended to it. Likewise for the -msan
	flag. Using a -buildmode option that requires non-default compile flags
	has a similar effect.
-ldflags 'flag list'
	arguments to pass on each go tool link invocation.
```

Since we are building go only version(`CGO_ENABLED=0`), we also need to build packages that we depend with cgo disabled, so `-a` (rebuild even if up to date) and to not mix with previously build version of the same package `installsuffix cgo` (use cgo suffix in installation directory).

[Linker options](https://golang.org/cmd/link/) are used to make binary small by striping unnecessary symbols.
```
-s
	Omit the symbol table and debug information.
-w
	Omit the DWARF symbol table.
``` 

On the other hand, this FAQ might not be place to explain these options. It might be better to use bare `go build` and using alpine based image for both stages.

Reply:
If we are going to include all those flags, quick comments or links to explain the magic are required IMO. If there is a simpler example that only demonstrates the dep aspect, then that would be preferable.

Reply:
Decided to only use the -o flag, and I added a short comment why.